subsys/mgmt/fmfu/src/mgmt_fmfu.c

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <logging/log.h>
+#include <mgmt/mgmt.h>
+#include <nrf_fmfu.h>
+#include <mgmt/mgmt_fmfu.h>
+#include "cborattr/cborattr.h"
+#include <stats/stats.h>
+
+LOG_MODULE_REGISTER(mgmt_fmfu, CONFIG_MGMT_FMFU_LOG_LEVEL);
+
+/* Define full modem update mgmt spesific SMP return codes */
+#define MGMT_ERR_EMODEM_INVALID_COMMAND 200
+#define MGMT_ERR_EMODEM_FAULT           201
+
+/* Define full modem mgmt commands */
+#define FMFU_MGMT_ID_GET_HASH 0
+#define FMFU_MGMT_ID_UPLOAD 1
+
+/* Define full modem IPC failures*/
+#define NRF_FMFU_RET_COMMAND_FAILED -3
+#define NRF_FMFU_RET_COMMAND_FAULT -4
+
+#define FIRMWARE_SHA_LENGTH 32
+
+/* Struct used for parsing upload requests. */
+struct firmware_packet {
+	uint8_t data[SMP_PACKET_MTU];
+	uint32_t target_address;
+	uint32_t data_len;
+};
+
+static int unpack(struct mgmt_ctxt *ctxt, struct firmware_packet *packet,
+		  bool *whole_file_received)
+{
+	static uint32_t file_length;
+	unsigned long long offset;
+	unsigned long long read_file_length;
+	size_t data_len;
+	size_t firmware_sha_len;
+	uint8_t firmware_sha[FIRMWARE_SHA_LENGTH];
+
+	CborError cbor_err = 0;
+	unsigned long long fw_target_address = 0;
+	/* Description/skeleton of CBOR object */
+	const struct cbor_attr_t off_attr[] = {
+		[0] = {
+			.attribute = "data",
+			.type = CborAttrByteStringType,
+			.addr.bytestring.data = packet->data,
+			.addr.bytestring.len = &data_len,
+			.len = SMP_PACKET_MTU
+		},
+		[1] = {
+			.attribute = "len",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &read_file_length,
+			.nodefault = true
+		},
+		[2] = {
+			.attribute = "off",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &offset,
+			.nodefault = true
+		},
+		/* Unused but needed for parsing the CBOR*/
+		[3] = {
+			.attribute = "sha",
+			.type = CborAttrByteStringType,
+			.addr.bytestring.data = firmware_sha,
+			.addr.bytestring.len = &firmware_sha_len,
+			.len = FIRMWARE_SHA_LENGTH,
+		},
+		[4] = {
+			.attribute = "addr",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &fw_target_address,
+			.nodefault = true
+		},
+		/* End of array */
+		[6] = { 0 },
+	};
+
+
+	cbor_err = cbor_read_object(&ctxt->it, off_attr);
+	if (cbor_err) {
+		return -MGMT_ERR_EINVAL;
+	}
+
+	if (offset == 0) {
+		file_length = (uint32_t)read_file_length;
+	}
+
+	packet->target_address  = (uint32_t)fw_target_address;
+	packet->data_len = (uint32_t)data_len;
+	uint32_t new_offset = (uint32_t)offset + packet->data_len;
+	*whole_file_received = (((uint32_t)offset + packet->data_len) == file_length);
+	LOG_INF("Off+pk_len: 0x%x, file_length:%x", new_offset, file_length);
```

Comment:
Should this be DBG? Do we get this for every packet?

Reply:
Added.